File: source/client/web/invoice-book-web/public_html/js/aatl-ib/service/employee-service.js
/* 
 * Copyright (c) 2018 Absolute Apogee Technologies Ltd. All rights reserved.
 * 
 * =============================================================================
 * Revision History:
 * Date         Author          Detail
 * -----------  --------------  ------------------------------------------------
 * 2018-Dec-25  GShokar         Created
 * =============================================================================
 */

"use strict";

$aatl_ib.EmployeeService = {

    find: function (criteria, callback) {

        $aatl_ib.ApiService.get("employees/find",
                criteria,
                function (res, err) {
                    if (err) {
                        callback(null, {messages: ["Faild to get employee search results: " + err]});

                    } else if (res.status === "failure") {
                        callback(null, {messages: [res.message]});
                    } else if (res.status === "success") {
                        let employees = JSON.parse(res.data);
                        callback(employees);
                    } else {
                        callback(null, {messages: ["Invalid response from server"]});
                    }
                });
    },
    list: function (callback) {

        $aatl_ib.ApiService.get("employees",
        {},
                function (res, err) {
                    if (err) {
                        callback(null, {messages: ["Faild to get employee list: " + err]});

                    } else if (res.status === "failure") {
                        callback(null, {messages: [res.message]});
                    } else if (res.status === "success") {
                        let clients = JSON.parse(res.data);
                        callback(clients);
                    } else {
                        callback(null, {messages: ["Invalid response from server"]});
                    }
                });
    },

    get: function (number, callback) {

        if (number === undefined || number === null) {
            callback({
                number: "",
                lastName: "",
                firstName: "",
                birthDate: "",
                address: {
                    address1: "",
                    address2: "",
                    city: "",
                    province: $aatl_ib.LookupService.defaultProvince.code,
                    postalCode: ""
                },
                contact: {
                    phone: "",
                    email: ""
                }
            });
        } else {
            $aatl_ib.ApiService.get("employees/" + number,
                    {},
                    function (res, err) {
                        if (err) {
                            callback(null, {messages: ["Faild to get employee detail: " + err]});

                        } else if (res.status === "failure") {
                            callback(null, {messages: [res.message]});
                        } else if (res.status === "success") {
                            let employee = JSON.parse(res.data);
                            callback(employee);
                        } else {
                            callback(null, {messages: ["Invalid response from server"]});
                        }
                    });
        }
    },

    save: function (employee, callback) {

        let err = $aatl_ib.EmployeeService.validate(employee);

        if (err.messages.length === 0) {
            $aatl_ib.ApiService.put("employees",
                    employee,
                    function (res, serverErr) {
                        if (serverErr) {

                            err.messages.push("Failed to save the employee: " + serverErr);
                            callback(employee, err);

                        } else if (res.status === "failure") {

                            err.messages.push(res.message);
                            callback(employee, err);

                        } else if (res.status === "success") {

                            if (Array.isArray(res.warningMessages) && res.warningMessages.length > 0) {
                                err.messages = res.warningMessages;
                            }

                            callback(JSON.parse(res.data), err);
                        } else {
                            callback(employee, "Invalid response from server");
                        }
                    });

            return;
        }

        callback(employee, err);
    },

    validate: function (employee) {

        let err = {messages: []};

        if (employee) {

            if ($aatl_ib.utils.isStringEmpty(employee.lastName)) {
                err.messages.push("Please enter employee last name.");
            }

            if ($aatl_ib.utils.isStringEmpty(employee.firstName)) {
                err.messages.push("Please enter employee first name.");
            }

            if ($aatl_ib.utils.isStringEmpty(employee.birthDate)) {
                err.messages.push("Please enter employee date of birth.");
            }

            if (employee.address && $aatl_ib.utils.isStringEmpty(employee.address.address1)) {
                err.messages.push("Please enter address line 1.");
            }

            if (employee.address && $aatl_ib.utils.isStringEmpty(employee.address.city)) {
                err.messages.push("Please enter city.");
            }

            if (employee.address && $aatl_ib.utils.isStringEmpty(employee.address.province)) {
                err.messages.push("Please enter province.");
            }

            if (employee.address && $aatl_ib.utils.isStringEmpty(employee.address.postalCode)) {
                err.messages.push("Please enter postal code.");
            }
        }

        return err;
    }
};

File: source/bl/invoice-book-service/src/main/java/ca/aatl/app/invoicebook/bl/ejb/EmployeeService.java
/*
 * Copyright (c) 2018 Absolute Apogee Technologies Ltd. All rights reserved.
 * 
 * =============================================================================
 * Revision History:
 * Date         Author          Detail
 * -----------  --------------  ------------------------------------------------
 * 2018-Dec-25  GShokar         Created
 * =============================================================================
 */
package ca.aatl.app.invoicebook.bl.ejb;

import ca.aatl.app.invoicebook.data.jpa.dao.EmployeeDao;
import ca.aatl.app.invoicebook.data.jpa.entity.Address;
import ca.aatl.app.invoicebook.data.jpa.entity.Contact;
import ca.aatl.app.invoicebook.data.jpa.entity.Employee;
import ca.aatl.app.invoicebook.exception.DataValidationException;
import ca.aatl.app.invoicebook.util.AppUtils;
import java.util.List;
import javax.ejb.EJB;
import javax.ejb.LocalBean;
import javax.ejb.Stateless;

/**
 *
 * @author GShokar
 */
@Stateless
@LocalBean
public class EmployeeService {

    @EJB
    EmployeeDao dao;

    @EJB
    LookupService lookupService;

    @EJB
    SequenceService sequenceService;

    public List<Employee> find(String name, String phone) throws Exception {
        return dao.find(name, phone);
    }

    public Employee find(String number) throws Exception {

        return dao.find(number);
    }

    public Employee newEntity() throws Exception {

        Employee employee = new Employee();

        employee.setActive(true);

        employee.setAddress(new Address());

        employee.getAddress().setCountry(lookupService.defaultCountry());
        employee.getAddress().setProvince(lookupService.defaultProvince());

        employee.setContact(new Contact());

        return employee;
    }

    public void save(Employee entity) throws Exception {

        beforeSave(entity);

        dao.save(entity);
    }

    private void beforeSave(Employee entity) throws DataValidationException {

        validate(entity);

        if (entity.getId() == null) {
            entity.setAddedBy(entity.getLastUpdatedBy());
            entity.setAddedDate(entity.getLastUpdatedDate());
            entity.getGuid();

            synchronized (this) {
                entity.setNumber(sequenceService.next("EmployeeNumber"));
            }
        }

        if (entity.getAddress() != null && entity.getAddress().getId() == null) {

            entity.getAddress().setAddedBy(entity.getLastUpdatedBy());
            entity.getAddress().setAddedDate(entity.getLastUpdatedDate());
            entity.getAddress().setLastUpdatedBy(entity.getLastUpdatedBy());
            entity.getAddress().setLastUpdatedDate(entity.getLastUpdatedDate());
            entity.getAddress().getGuid();
        }

        if (entity.getContact() != null && entity.getContact().getId() == null) {
            entity.getContact().setAddedBy(entity.getLastUpdatedBy());
            entity.getContact().setAddedDate(entity.getLastUpdatedDate());
            entity.getContact().setLastUpdatedBy(entity.getLastUpdatedBy());
            entity.getContact().setLastUpdatedDate(entity.getLastUpdatedDate());
            entity.getContact().getGuid();
        }
    }

    public void validate(Employee entity) throws DataValidationException {

        boolean rtnValue = true;
        boolean checkDuplicate = true;

        StringBuilder sb = new StringBuilder();

        if (entity != null) {

            if (AppUtils.isNullOrEmpty(entity.getFirstName())) {
                rtnValue = false;
                checkDuplicate = false;
                sb.append(System.lineSeparator());
                sb.append("Enter the employee first name.");
            }

            if (AppUtils.isNullOrEmpty(entity.getLastName())) {
                rtnValue = false;
                checkDuplicate = false;
                sb.append(System.lineSeparator());
                sb.append("Enter the employee last name.");
            }

            if (entity.getBirthDate() == null) {
                rtnValue = false;
                checkDuplicate = false;
                sb.append(System.lineSeparator());
                sb.append("Enter the employee date of birth.");
            } else if (entity.getBirthDate().after(AppUtils.currentDate())) {
                rtnValue = false;
                checkDuplicate = false;
                sb.append(System.lineSeparator());
                sb.append("Employee's date of birth can't be in future.");
            }

            if (checkDuplicate && dao.isExists(entity.getId(), entity.getName(), entity.getBirthDate())) {
                rtnValue = false;
                sb.append(System.lineSeparator());
                sb.append("This employee already exists.");
            }

            if (entity.getAddress() != null) {

                Address address = entity.getAddress();

                if (AppUtils.isNullOrEmpty(address.getAddress1())) {
                    rtnValue = false;
                    sb.append(System.lineSeparator());
                    sb.append("Enter the address.");
                }

                if (AppUtils.isNullOrEmpty(address.getCity())) {
                    rtnValue = false;
                    sb.append(System.lineSeparator());
                    sb.append("Enter the city.");
                }

                if (AppUtils.isNullOrEmpty(address.getPostalCode())) {
                    rtnValue = false;
                    sb.append(System.lineSeparator());
                    sb.append("Enter the postal code.");
                }
            }

            if (entity.getContact() != null) {
                Contact contact = entity.getContact();

                if (AppUtils.isNullOrEmpty(contact.getPhone())) {
                    rtnValue = false;
                    sb.append(System.lineSeparator());
                    sb.append("Enter the phone number.");
                }

            }
        }

        if (!rtnValue) {
            throw new DataValidationException(sb.toString());
        }

    }

    public List<Employee> list() throws Exception{
        return dao.list();
    }
}
